### PR TITLE
[TG-9450] Allocate reference arrays with constant size in assignments from json

### PR DESCRIPTION
**Pull request by romainbrenguier**
This is to ensure static reference arrays retrieved from a json file are assigned with a constant size (when not marked as `@nondetLength`). This is so that array-cell-sensitivity can apply to such arrays.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: d114b94384365793722503b5bc046f363001985b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/128489226

File: jbmc/src/java_bytecode/assignments_from_json.cpp [hunk 1]
```diff
@@ -28,9 +28,6 @@ Author: Diffblue Ltd.
 /// ones that are not marked `const` may be mutated.
 struct object_creation_infot
```
Comment (smowton) [inline jbmc/src/java_bytecode/assignments_from_json.cpp:29]: Commit message: "We make is to make it so" -> "We make it so"

File: jbmc/src/java_bytecode/code_with_references.h [hunk 2]
```diff
@@ -33,4 +33,72 @@ struct object_creation_referencet
...
+};
+
+/// Wrapper around a list of shared pointer to code_with_referencest objects,
+/// which provides a nicer interface.
+class code_with_references_listt
+{
+public:
+  std::list<std::shared_ptr<code_with_referencest>> list;
+
+  void add(code_without_referencest code);
```
Comment (smowton) [inline jbmc/src/java_bytecode/code_with_references.h:97]: You seem to use `add(code_without_referencest{...})` a lot, so perhaps just create an `add(codet)` overload?

File: jbmc/src/java_bytecode/assignments_from_json.cpp [hunk 3]
```diff
@@ -755,12 +755,7 @@ static get_or_create_reference_resultt get_or_create_reference(
       reference.array_length =
         info.allocate_objects.allocate_automatic_local_object(
           java_int_type(), "user_specified_array_length");
-      code.add(code_without_referencest{code_assignt{
-        *reference.array_length, side_effect_expr_nondett{java_int_type()}}});
-      code.add(code_without_referencest{code_assumet{binary_predicate_exprt{
-        *reference.array_length, ID_ge, from_integer(0, java_int_type())}}});
-      code.add(code_without_referencest{
-        allocate_array(reference.expr, *reference.array_length, info.loc)});
+      code.add(reference_allocationt{id, info.loc});
```
Comment (smowton) [inline jbmc/src/java_bytecode/assignments_from_json.cpp:748]: Commit "Add reference allocation code for references" doesn't seem to do what it says (it only deletes code from `get_or_create_reference`)

File: jbmc/src/java_bytecode/code_with_references.cpp [hunk 4]
```diff
@@ -63,3 +63,9 @@ void code_with_references_listt::append(code_with_references_listt &&other)
 {
   list.splice(list.end(), other.list);
 }
+
+void code_with_references_listt::add_to_front(code_without_referencest code)
```
Comment (smowton) [inline jbmc/src/java_bytecode/code_with_references.cpp:72]: Commit message: typo "begginning"

**Review by smowton — APPROVED** [review]
I don't fully grasp why the references are necessary (perhaps add some high-level docs about what they're doing?), but I don't see anything troubling.

**Comment by romainbrenguier** [comment]
@smowton 
> I don't fully grasp why the references are necessary (perhaps add some high-level docs about what they're doing?), but I don't see anything troubling.

This is already explained in the documentation of `assign_from_json`, we need them for objects that are reference-equal. For instance `Object a = new Object(); Object b = new Object(); Object c = a;" We have `c == a` but not `b == a` and the way to distinguish the two cases in the json is to make `c` a reference to `a`.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 69348e00dc69c91cddd5d9e0c1d6d53523597953).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/128702645

File: jbmc/regression/jbmc/array-cell-sensitivity-static-fields/test-char-array-ref.desc [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+Generated 0 VCC[(]s[)], 0 remaining after simplification
+assertion at file Test.java line 29 .*: SUCCESS
+^EXIT=0$
+^SIGNAL=0$
+--
+--
+Not that the json file has been modified on purpose so that this checks
+that it is actually loaded and not retrieved by execution of the Util function.
+This test does not work with symex-driven-lazy-loading because this option is
+not compatible with --property.
```
Comment (majakusber) [inline jbmc/regression/jbmc/array-cell-sensitivity-static-fields/test-char-array-ref.desc]: :pick: Missing new line

File: jbmc/src/java_bytecode/assignments_from_json.cpp [hunk 6]
```diff
@@ -902,11 +902,10 @@ code_with_references_listt assign_from_json_rec(
   return result;
 }
 
-void assign_from_json(
+code_with_references_listt assign_from_json(
```
Comment (majakusber) [inline jbmc/src/java_bytecode/assignments_from_json.cpp:905]: :pick: Typo in the commit message: `the all the value pairs` -> `all the value pairs`

**Review by majakusber — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: bc47e4906eaef342bbfc10604080e532f2c391b1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/128757153

**Pull request merged**